man/optiups.8
.TH OPTIUPS 8 "Fri Jan 20 2006" "" "Network UPS Tools (NUT)" 
.SH NAME  
optiups \- Driver for Opti-UPS (Viewsonic) UPS and Zinto D (ONLINE-USV) equipment
.SH NOTE
This man page only documents the hardware\(hyspecific features of the
optiups driver.  For information about the core driver, see  
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
optiups was originally written against a PowerES 280es in nut-0.45.  It was
revised for nut 2.0.1 and tested against a PowerES 420E.  It is expected to
work with at least the PowerES, PowerPS, and PowerVS models.

This driver additionally supports a Zinto D from ONLINE USV-Systeme AG
because of their very similar commands, but it's unknown if it also works
with other ups from them.

This driver will not work with the PowerES stock serial cable.  You will need
to construct your own three conductor cable:

.RS
UPS 6 -> PC 3
.RE
.RS
UPS 9 -> PC 2
.RE
.RS
UPS 4 -> PC 5
.RE

The cable for Online-USV uses pin UPS 7 (not UPS 4) -> PC 5.

.SH EXTRA ARGUMENTS

This driver supports the following optional settings in the
\fBups.conf\fR(5):

.IP "status_only"

Only poll for critical status information.  Without this, optiups (and
all nut drivers) poll all sorts of information from the UPS fairly often.
It is probably not often enough to hurt anything, so this option probably
is not very useful, unless you have a flaky serial connection or a highly
loaded machine.

.IP "nowarn_noimp"

Does not print warnings when the UPS reports that a variable is not
implemented or not pollable.  Without the option you will get a message
sent to your system logs each time nut polls the ups.  If you specify
\fBnowarn_noimp\fR, this message will only be logged once.

.IP "fake_lowbatt"

This forces the low battery flag true.  Without it, if you want to test your
UPS, you will have to unplug it and wait until the battery drops to a low/critical
voltage level before nut will respond and power down your system.  With the flag,
nut should power down the system soon after you pull the plug.  When you are done
testing, you should remove this flag.

.IP "powerup"

Zinto D from ONLINE-USV cannot be identified when switched to standby.
Set this flag to allow the driver to power-up your Zinto UPS.
This will also power-up your equipment connected to the ups!

.SH BUGS

On the 420E, ups.serial and ups.temperature are unsupported features.  This
is not a bug in nut or the nut driver, just the way things are with this UPS.

.SH AUTHOR
Russell Kroll, Scott Heavner, Matthias Goebl

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.5 - what's new since 2.0.4:

 This release is a backport of the development version.  Many changes
 have already been backported previously.  Thus it is more a
 synchronisation release, though it includes many bugfixes and support
 for new models.

 - support for new devices: APC Smart-UPS with 6TI firmware; Belkin
   Small Enterprise F6C1500-TW-RK; Compaq R3000 XR, R5500 XR; Cyber
   Power 550SL, 725SL, 685AVR, 800AVR, 1200AVR, AE550, BC1200 and
   many similar models; Eltek; Inform GUARD; Microsol Rhino; Opti-UPS
   PowerES 420E; PowerMan RealSmart, BackPro; Powerware PW9315 3-phase;
   SOLA 305; Tripp-Lite SMART550USB, SMART2200RMXL2U, OMNI1000LCD,
   OMNI900LCD, OMNI650LCD, 1500 LCD, AVR550U; Viewsonic PowerES 420E.

 - bcmxcp: added 3-phase support

 - megatec: better hardware support, more instant commands

 - mge-hid: support more instant commands

 - newhidups: fixed APC and Tripp Lite bugs, various memory bugs,
   improved report buffering, improved Solaris support, added
   '-x explore' option for easy diagnosis of new devices

 - solis: shutdown programming, support new cables, Solaris support

 - tripplite_usb: updated SMARTPRO support, fixed OL/OB reporting, 
   better error handling, some memory bugs

 - new dummy-ups driver simulator

 - added HTML interface for access to CGI scripts

 - powerpanel: new driver for CyberPower BC1200 and similar models
 
 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.0.4 - what's new since 2.0.3:

 - The newhidups critical bug (segmentation fault) has been fixed. It has
   also received some more care, like buxfixes and new models support and
   enhancement for Solaris.
   [Peter Selinger and Arnaud Quette]

 - A bug has been fixed in NUT core to support resuming from suspend-to-disk.
   This should also fix other similar issues, like time synchronisation
   through the NTP - Network Time Protocol.
   [Arjen de Korte]

 - The mge-shut driver now better detects the Low Battery status, support
   new models and fixes some wrong status and data. It also fixes some
   issue where the UPS wasn't restarting (refer to mge-shut manpage).
   [Arnaud Quette]

 - The genericups custom configuration through ups.conf is working again
   [Arjen de Korte]

 - The genericups driver type 22 also support CyberPower 725SL
   (and maybe others SL models)
   [David Kaufman]

 - The new megatec driver, which will replace a bunch of drivers by nut 2.2
   (refer to docs/megatec.txt and UPGRADING) has been backported from the
   trunk (Development tree). The powermust driver has also received some
   attention.
   [Carlos Rodrigues]

 - The new rhino driver was added to support Microsol Rhino UPS hardware
   The solis has also been improved for solaris compatibility, and
   internal / external shutdown programming. solis can now save external
   shutdown programming to ups, and support new cables for solis 3
   [Silvino B. MagalhÃ£es]

 - Several fixes and improvements have been made to upsrw, upsset,
   cpsups, tripplite_usb and the FAQ.
   [Arjen de Korte and Charles Lepple]

---------------------------------------------------------------------------
Release notes for NUT 2.0.3 - what's new since 2.0.2:

 - The recent and major newhidups changes have been backported from the
   Development tree. It now:
   - supports models from MGE UPS SYSTEMS, APC and Belkin. Mustek and Unitek
     units are also recognized for development purpose,
   - handles better device reopening, after a disconnection,
   - handles multiple devices, with several parameters to find the right UPS.
   [Peter Selinger, Charles Lepple and Arnaud Quette]

 - The bcmxcp_usb driver has been added to support Powerware USB units.
   [Wolfgang Ocker and Kjell Claesson]

 - The tripplite_usb driver has been added to support Tripp Lite USB units.
   [Charles Lepple]

 - The sec driver is back as gamatronic
   [Gamatronic, Nadav Moskovitch]

 - The genericups driver has received official care from Gamatronic
   to add support for the Gamatronic UPS with alarm interface.
   [Gamatronic, Nadav Moskovitch]

 - The powermust driver now supports Soyntec Sekury C 500 and C 800 units.
   [Hanno Borns]

 - The mge-shut driver has received a bit of attention too, and enhance
   ups.model retrieval for some specific case (release 0.65)

 - The drivers don't change to the "statepath" directory anymore at
   initialisation time if called using -k. This avoid unneeded
   failure to poweroff the UPS if /var is already unmounted.
   [Gaspar Bakos]

 - The belkinunv driver now supports Belkin F6C1100-UNV
   [Dave Breiland]

 - The isbmex driver has been upgraded to version 0.05, which fixes
   various errors in formulas, add shutdown capability and revert
   back baudrate to B9600 (instead of B2400), as it broke the
   communication
   [Ricardo Martinezgarza]

 - The support of Sysgration UPGUARDS Pro650 in fentonups has
   been fixed
   [Simon J. Rowe]

 - The packaging files for Red Hat have received various fixes
   [Thomas Jarosch]

 - The solis driver has been fixed to avoid a naming colision and
   compile on Solaris
   [Paweł Kierdelewicz]

 - The snmp-ups driver has corrected the problem when exposing
   certain time data.

---------------------------------------------------------------------------
Release notes for NUT 2.0.2 - what's new since 2.0.1:

 - the newhidups USB driver has been improved a lot and is no more
   experimental. It also now has a basic APC support, which will
   soon replace the legacy hidups driver.

 - The mge-utalk driver has improved its support for old units.

 - The mge-shut driver has been improved for restart/shutdown
   sequences which was previously blocking the serial port.


drivers/optiups.c
/* optiups.c - OptiSafe UPS (very loosely based on the nut 0.45.5 driver)

   Copyright (C) 1999  Russell Kroll <rkroll@exploits.org>
   Copyright (C) 2006  Scott Heavner [Use my alioth acct: sheavner]

   Support for Zinto D from ONLINE USV (only minor differences to OptiSafe UPS)
   added by Matthias Goebl <matthias.goebl@goebl.net>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

   ==================================================================================
*/

#define HELP                                                                    "\n" \
   "**********************************************************"                 "\n" \
   "This driver has been tested only with the OPTI SAFE 420E using"             "\n" \
   "the custom cable described in the NUT Opti-UPS protocol page."              "\n" \
   "Seriously, it does _NOT_ work with the [windows] cable provided with"       "\n" \
   "the UPS or a standard serial cable. (I don't know if OPTI makes other"      "\n" \
   "UPS models that will work with another cable?)  Standard linux serial"      "\n" \
   "port drivers do not support the required DTR handshaking."                  "\n" \
   ""                                                                           "\n" \
   "     UPS 6 -> PC 3          This 3 wire cable pinout maps DTR to CTS."      "\n" \
   "     UPS 9 -> PC 2"                                                         "\n" \
   "     UPS 4 -> PC 5"                                                         "\n" \
   ""                                                                           "\n" \
   "This driver has also been tested with a Zinto D from Online-USV AG,"        "\n" \
   "using their special cable:"                                                 "\n" \
   "     UPS 6 -> PC 3"                                                         "\n" \
   "     UPS 9 -> PC 2"                                                         "\n" \
   "     UPS 7 -> PC 5"                                                         "\n" \
   "It works even with a pl2303 usb-serial converter."                          "\n" \
   "**********************************************************"                 "\n"

/* See http://www.networkupstools.org/protocols/optiups.html and the end of this
 * file for more information on the cable and the OPTI-UPS serial protocol used on
 * at least the older OPTI UPS models (420E, 820ES).
 */

#include "main.h"
#include "serial.h"
#include "optiups.h"

#define ENDCHAR		'\n'
#define IGNCHARS	"\r"

/* Our custom options available with -x */
#define OPTI_MINPOLL		"status_only"
#define OPTI_FAKELOW		"fake_lowbatt"
#define OPTI_NOWARN_NOIMP	"nowarn_noimp"
#define OPTI_POWERUP		"powerup"

/* All serial commands put their response in the same buffer space */
static char _buf[256];

/* Model */
static int optimodel = 0;
enum {
	OPTIMODEL_DEFAULT = 0,
	OPTIMODEL_ZINTO =1
};


/* Status bits returned by the "AG" command */
enum {
	OPTISBIT_NOOUTPUT = 2,
	OPTISBIT_OVERLOAD = 8,
	OPTISBIT_REPLACE_BATTERY = 16,
	OPTISBIT_ON_BATTERY_POWER = 32,
	OPTISBIT_LOW_BATTERY = 64
};

/* Helper struct for the optifill() function */
typedef struct ezfill {
	const char *cmd;
	const char *var;
	const float scale;
} ezfill;

/* These can be polled right into a string usable by NUT.
 * Others such as "AG" and "BV" require some transformation of the return value */
static ezfill _pollv[] = {
	{ "NV", "input.voltage" },
	{ "OL", "ups.load", 1.0 },
	{ "OV", "output.voltage" },
	{ "FF", "input.frequency", 0.1 },
	{ "BT", "ups.temperature" },
};
static ezfill _pollv_zinto[] = {
	{ "NV", "input.voltage", 2.0 },
	{ "OL", "ups.load", 1.0 },
	{ "OV", "output.voltage", 2.0 },
	{ "OF", "output.frequency", 0.1 },
	{ "NF", "input.frequency", 0.1 },
	{ "BT", "ups.temperature" },
};

/* model "IO" is parsed differently in upsdrv_initinfo() */
static ezfill _initv[] = {
	{ "IM", "ups.mfr" },
	{ "IZ", "ups.serial" },
	{ "IS", "ups.firmware" },
};

/* All serial reads of the OPTI-UPS go through here.  We always expect a CR/LF terminated
 *   response.  Unknown/Unimplemented commands return ^U (0x15).  Actions that complete
 *   successfully return ^F (0x06). */
static inline int optireadline()
{
	int r;
	usleep(150000);
	r = ser_get_line(upsfd, _buf, sizeof(_buf), ENDCHAR, IGNCHARS, 0, 500000 );
	_buf[sizeof(_buf)-1] = 0;
	if ( r > 0 )
	{
		if ( r < (int)sizeof(_buf) )
			_buf[r] = 0;
		if ( _buf[0] == 0x15 )
		{
			r=-2;
			upsdebugx(1, "READ: <unsupported command>");
		}
		if ( _buf[0] == 0x06 )
		{
			upsdebugx(2, "READ: <command done>");
		}
		else
		{
			upsdebugx(2, "READ: \"%s\"", _buf );
		}
	}
	else
		upsdebugx(1, "READ ERROR: %d", r );
	return r;
}

/* Send a command and read the response.  Command response is in global _buf.
 *   Return
 *      > 0 implies success.
 *      -1  serial timeout
 *      -2  unknown/unimplemented command
 */
static inline int optiquery( const char *cmd )
{
	upsdebugx(2, "SEND: \"%s\"", cmd );
	ser_send( upsfd, cmd );
	if ( optimodel == OPTIMODEL_ZINTO )
		ser_send( upsfd, "\r\n" );
	return optireadline();
}

/* Uses the ezfill structure to map UPS commands to the NUT variable destinations */
static void optifill( ezfill *a, int len )
{
	int i, r;

	/* Some things are easy to poll and store */
	for ( i=0; i<len; ++i )
	{
		if ( ( a[i].cmd == NULL ) || ( a[i].cmd[0] == 0 ) )
			continue;

		r = optiquery( a[i].cmd );
		if ( r < 1 )
		{
			if ( r == -2 )
			{
				if ( ! testvar(OPTI_NOWARN_NOIMP) )
					upslogx( LOG_NOTICE,
						"disabling poll of unsupported var %s",
						a[i].var );
				dstate_setinfo( a[i].var, "N/A" );
				a[i].cmd = NULL;
			}
			else
				upslogx( LOG_WARNING, "can't retrieve %s", a[i].var );
			continue;
		}
		if ( a[i].scale > 1e-20 )
		{
			float f = strtol( _buf, NULL, 10 ) * a[i].scale;
			dstate_setinfo( a[i].var, "%.1f", f );
		}
		else
		{
			dstate_setinfo( a[i].var, _buf);
		}
	}
}

/* Handle custom (but standardized) NUT commands */
static int instcmd(const char *cmdname, const char *extra)
{
        if (!strcasecmp(cmdname, "test.failure.start"))
	{
		optiquery( "Ts" );
                return STAT_INSTCMD_HANDLED;
        }
	else if (!strcasecmp(cmdname, "load.off"))
	{
		/* You do realize this will kill power to ourself.  Would probably only
		 *   be useful for killing power for a slave computer */
		if ( optimodel == OPTIMODEL_ZINTO )
		{
			optiquery( "Ct1" );
			optiquery( "Cs0000000" );
			sleep(2);
			return STAT_INSTCMD_HANDLED;
		}
		optiquery( "Ct0" );
		optiquery( "Cs00000000" );
                return STAT_INSTCMD_HANDLED;
        }
	else if (!strcasecmp(cmdname, "load.on"))
	{
		if ( optimodel == OPTIMODEL_ZINTO )
		{
			optiquery( "Ct1" );
			optiquery( "Cu0000000" );
			sleep(2);
			return STAT_INSTCMD_HANDLED;
		}
		optiquery( "Ct0" );
		optiquery( "Cu00000000" );
                return STAT_INSTCMD_HANDLED;
        }
	else if (!strcasecmp(cmdname, "shutdown.return"))
	{
		/* This shuts down the UPS.  When the power returns to the UPS,
		 *   it will power back up in its default state. */
		if ( optimodel == OPTIMODEL_ZINTO )
		{
			optiquery( "Ct1" );
			optiquery( "Cu0000010" );
			optiquery( "Cs0000001" );
			return STAT_INSTCMD_HANDLED;
		}
		optiquery( "Ct1" );
		optiquery( "Cs00000010" );
                return STAT_INSTCMD_HANDLED;
        }
	else if (!strcasecmp(cmdname, "shutdown.stayoff"))
	{
		/* This actually stays off as long as the batteries hold,
		 *   if the line power comes back before the batteries die,
		 *   the UPS will never powerup its output stage!!! */
		if ( optimodel == OPTIMODEL_ZINTO )
		{
			optiquery( "Ct1" );
			optiquery( "Cs0000001" );
			return STAT_INSTCMD_HANDLED;
		}
		optiquery( "Ct0" );
		optiquery( "Cs00000010" );
                return STAT_INSTCMD_HANDLED;
        }
	else if (!strcasecmp(cmdname, "shutdown.stop"))
	{
		/* Aborts a shutdown that is couting down via the Cs command */
		optiquery( "Cs-0000001" );
                return STAT_INSTCMD_HANDLED;
        }

        upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
        return STAT_INSTCMD_UNKNOWN;
}

/* Handle variable setting */
static int setvar(const char *varname, const char *val)
{
	int status;

	if (sscanf(val, "%d", &status) != 1) {
		return STAT_SET_UNKNOWN;
	}

	if (strcasecmp(varname, "outlet.1.switch") == 0) {    
		status = status==1 ? 1 : 0;
		dstate_setinfo( "outlet.1.switch", "%d", status);
		optiquery(status ? "Oi11" : "Oi10");
		dstate_dataok();
		return STAT_SET_HANDLED;
	}

	return STAT_SET_UNKNOWN;
}

void upsdrv_initinfo(void)
{
	int r;

	dstate_setinfo("driver.version.internal", "%s", DRV_VERSION);

	/* If an Zinto Online-USV is off, switch it on first. */
	/* It sends only "2" when off, without "\r\n", and doesn't */
	/* answer other commands. Therefore without power we'll be */
	/* unable to identify the ups. */
	if ( testvar(OPTI_POWERUP) && optiquery( "AG" ) < 1 )
	{
		ser_send( upsfd, "AG\r\n" );
		r = ser_get_char(upsfd, &_buf[0], 1, 0);
		if ( r == 1 && _buf[0] == '2' )
		{
			upslogx( LOG_WARNING, "ups was off, switching it on" );
			optiquery( "Ct1" );
			optiquery( "Cu0000000" );
			/* wait for power up */
			sleep(15);
		}
	}

	/* Autodetect an Online-USV (only Zinto D is known to work) */
	r = optiquery( "IM" );
	if ( r > 0 && !strcasecmp(_buf, "ONLINE") )
	{
		optimodel = OPTIMODEL_ZINTO;
		optiquery( "Om11" );
		optiquery( "Om21" );
		optiquery( "ON" );
	}

	optifill( _initv, sizeof(_initv)/sizeof(_initv[0]) );

	/* Parse out model into longer string -- is this really USEFUL??? */
	r = optiquery( "IO" );
	if ( r < 1 ) 
		fatal_with_errno( "can't retrieve model" );
	else
	{
		switch ( _buf[r-1] )
		{
			case 'E':
			case 'P':
			case 'V':
				dstate_setinfo("ups.model", "Power%cS %s", _buf[r-1], _buf );
				break;
			default:
				dstate_setinfo("ups.model", "%s", _buf );
				break;
		}
	}

	/* Parse out model into longer string */
	r = optiquery( "IM" );
	if ( r > 0 && !strcasecmp(_buf, "ONLINE") )
	{
		dstate_setinfo("ups.mfr", "ONLINE USV-Systeme AG");
		r = optiquery( "IO" );
		if ( r < 1 ) 
			fatal_with_errno( "can't retrieve model" );
		switch ( _buf[0] )
		{
			case 'D':
				dstate_setinfo("ups.model", "Zinto %s", _buf );
				break;
			default:
				dstate_setinfo("ups.model", "%s", _buf );
				break;
		}
	}

        dstate_addcmd("test.failure.start");
        dstate_addcmd("load.off");
        dstate_addcmd("load.on");
        if( optimodel != OPTIMODEL_ZINTO )
        	dstate_addcmd("shutdown.stop");
        dstate_addcmd("shutdown.return");
        dstate_addcmd("shutdown.stayoff");
	upsh.instcmd = instcmd;

	if ( optimodel == OPTIMODEL_ZINTO )
	{
		dstate_setinfo("outlet.0.desc", "%s", "Main Outlet 1+2");
		dstate_setinfo("outlet.1.desc", "%s", "Switchable Outlet 3+4");
		dstate_setinfo("outlet.0.id", "%d", 1);
		dstate_setinfo("outlet.1.id", "%d", 2);
		dstate_setinfo("outlet.0.switchable", "%d", 0);
		dstate_setinfo("outlet.1.switchable", "%d", 1);
		dstate_setinfo("outlet.1.switch", "%d", 1);
		dstate_setflags("outlet.1.switch", ST_FLAG_RW | ST_FLAG_STRING);
		dstate_setaux("outlet.1.switch", 1);
		upsh.setvar = setvar;	
	}
}

void upsdrv_updateinfo(void)
{
	int r = optiquery( "AG" );

	/* Online-UPS send only "2" when off, without "\r\n" */
	if ( r < 1 && optimodel == OPTIMODEL_ZINTO )
	{
		ser_send( upsfd, "AG\r\n" );
		r = ser_get_char(upsfd, &_buf[0], 1, 0);
		if ( r == 1 && _buf[0] == '2' )
		{
			status_init();
			status_set("OFF");
			status_commit();
			return;
		}
	}

	if ( r < 1 )
	{
		upslogx(LOG_ERR, "can't retrieve ups status" );
		dstate_datastale();
	}
	else
	{
		int s = strtol( _buf, NULL, 16 );
		status_init();
		if ( s & OPTISBIT_OVERLOAD )
	 	 	status_set("OVER");
		if ( s & OPTISBIT_REPLACE_BATTERY )
	 	 	status_set("RB");
		if ( s & OPTISBIT_ON_BATTERY_POWER )
	 	 	status_set("OB");
		else
	 	 	status_set("OL");
		if ( s & OPTISBIT_NOOUTPUT )
			status_set("OFF");
		if ( s & OPTISBIT_LOW_BATTERY )
	 	 	status_set("LB");
		if ( testvar(OPTI_FAKELOW) )  /* FOR TESTING */
	 	 	status_set("LB");
	        status_commit();
	 	dstate_dataok();
	}

	/* Get out of here now if minimum polling is desired */
	if ( testvar(OPTI_MINPOLL) )
		return;

	/* read some easy settings */
	if ( optimodel == OPTIMODEL_ZINTO )
		optifill( _pollv_zinto, sizeof(_pollv_zinto)/sizeof(_pollv_zinto[0]) );
	else
		optifill( _pollv, sizeof(_pollv)/sizeof(_pollv[0]) );

	/* Battery voltage is harder */
	r = optiquery( "BV" );
	if ( r < 1 )
		upslogx( LOG_WARNING, "cannot retrieve battery voltage" );
	else
	{
		float p, v = strtol( _buf, NULL, 10 ) / 10.0;
		dstate_setinfo("battery.voltage", "%.1f", v );

		/* battery voltage range: 10.4 - 13.0 VDC */
		p = ((v  - 10.4) / 2.6) * 100.0;
		if ( p > 100.0 )
			p = 100.0;
		dstate_setinfo("battery.charge", "%.1f", p );
	}
}

void upsdrv_shutdown(void)
{
	/* OL: this must power cycle the load if possible */
	/* OB: the load must remain off until the power returns */

	/* If get no response, assume on battery & battery low */
	int s = OPTISBIT_ON_BATTERY_POWER | OPTISBIT_LOW_BATTERY;

	int r = optiquery( "AG" );
	if ( r < 1 )
	{
		upslogx(LOG_ERR, "can't retrieve ups status during shutdown" );
	}
	else
	{
		s = strtol( _buf, NULL, 16 );
	}

	/* Turn output stage back on if power returns - but really means
	 *   turn off ups if on battery */
	optiquery( "Ct1" );

	/* What happens, if the power comes back *after* reading the ups status and
	 * before the shutdown command? For "Online-UPS Zinto D" *always* asking for
	 * "shutdown shortly and power-up later" works perfectly, because it forces
	 * a power cycle, even for the named race condition.
	 * For Opti-UPS I have no information, so I wouldn't dare to change it.
	 * BTW, Zinto expects only 7 digits after Cu/Cs.
	 * (Matthias Goebl)
	 */
	if ( optimodel == OPTIMODEL_ZINTO )
	{
		/* On line power: Power up in 60 seconds (30 seconds after the following shutdown) */
		/* On battery: Power up when the line power returns */
		optiquery( "Cu0000060" );
		/* Shutdown in 30 seconds */
		optiquery( "Cs0000030" );
		return;
	}

	/* Just cycling power, schedule output stage to come back on in 60 seconds */
	if ( !(s&OPTISBIT_ON_BATTERY_POWER) )
		optiquery( "Cu00000600" );

	/* Shutdown in 8 seconds */
	optiquery( "Cs00000080" );
}

void upsdrv_help(void)
{
	printf(HELP);
}

/* list flags and values that you want to receive via -x */
void upsdrv_makevartable(void)
{
	addvar(VAR_FLAG, OPTI_MINPOLL, "Only poll for critical status variables");
	addvar(VAR_FLAG, OPTI_FAKELOW, "Fake a low battery status" );
	addvar(VAR_FLAG, OPTI_NOWARN_NOIMP, "Supress warnings of unsupported commands");
	addvar(VAR_FLAG, OPTI_POWERUP, "(Zinto D) Power-up UPS at start (cannot identify a powered-down Zinto D)");
}

void upsdrv_banner(void)
{
	printf("Network UPS Tools - Opti-UPS driver %s (%s)\n\n", 
		DRV_VERSION, UPS_VERSION);
}

void upsdrv_initups(void)
{
	upsfd = ser_open(device_path);
	ser_set_speed(upsfd, device_path, B2400);

	/* NO PARSED COMMAND LINE VARIABLES */
}

void upsdrv_cleanup(void)
{
	ser_close(upsfd, device_path);
}

/*******************************************
 * COMMANDS THAT QUERY THE UPS
 *******************************************
 * IM - manufacturer <ViewSonic>
 * IO - model <UPS-420E>
 * IS - firmware version <V1.2B>
 * IZ - serial number <> (unsupported on 420E, returns ^U)
 *
 * BS - ??? returns <2>
 * BV - battery voltage (in deciVolts) <0140>
 * BT - battery temperature (deg C) <0033> (returns ^U on OptiUPS 420E)
 *
 * NF - input line frequency <600>
 * NV - input line voltage <116>
 *
 * OS - ?? return <3>
 * OF - output stage frequency (in 0.1 Hz) <600>
 * OV - output stage voltage <118>
 * OL - output stage load <027>
 * 
 * FV - Input voltage <120>
 * FF - Input Frequency (0.1Hz) <600>
 * FO - Output volts <120>
 * FR - Output Frequency (0.1Hz) <600>
 * FA - Output VA <420>
 * FP - Ouptu power <252>
 * FU - ?? returns <2>
 * FB - ??
 * FH - High Transfer Point <144>
 * FL - Low Transfer Point <093>
 * FT - Transfer point? <121>
 *
 * AG - UPS status (bitmapped ASCII hex value) <00>
 *         bit 2: 1 = <set when output stage off?>
 *         bit 3: 1 = overload
 *         bit 4: 1 = replace battery
 *         bit 5: 1 = on battery, 0 = on line
 *         bit 6: 1 = low battery 
 * TR - Test results <00>
 *         00 = Unknown
 *         01 = Passed
 *         02 = Warning
 *         03 = Error
 *         04 = Aborted
 *         05 = In Progress
 *         06 = No test init 
 *
 *******************************************
 * ACTIONS
 *******************************************
 *
 * Ts - Start test
 * Ct0 - set power down mode (when running on battery)
 *         Ct0 = power down only ouput stage
 *         Ct1 = complete power down
 * Cs00000000 - power down after delay (count in 0.1s)
 *         Cs00000100 = power down in 10s
 *         Cs-0000001 = cancel power down request
 * Cu00000000 - power down after delay (count in 0.1s)
 *         Cu00000050 = power up in 5s
 *         Cu00000000 = power up now
 *
 * CT - returns last setting passed to Ct command <0>
 *
 *******************************************
 * UNKNOWN COMMANDS (on 420e anyways)
 *******************************************
 * Fu
 * Fv
 * Ff
 * Fo
 * Fr
 * Fb
 * Fl
 * Fh
 */
